dlt/common/json/__init__.py

```diff
@@ -126,6 +137,12 @@ def custom_pua_encode(obj: Any) -> str:
     elif isinstance(obj, Enum):
         # Enum value is just int or str
         return obj.value  # type: ignore[no-any-return]
+    elif _custom_encoder is not None:
+        ret = _custom_encoder(obj)
+        if ret is not None:
+            if not isinstance(ret, str):
+                raise TypeError("Custom encoder must return a string for PUA encoding")
```

Comment:
I did this because of the signature of `custom_pua_encode` which is more narrow than the one from `custom_encode`.
We can also:
* allow two separate encoders, one for `custom_pua_encode` and one for `custom_encode`
* not throw here and just document the requirement; I can see that the `Enum` case above just ignores a potential any return value and objects with `_asdict` extension are actually not returning `str` either - is it possible the signature of `custom_pua_encode` is outdated?